pkgs/by-name/vc/vcpkg-tool/package.nix

```
@@ -47,23 +49,100 @@ stdenv.mkDerivation (finalAttrs: {
     "-DVCPKG_DEPENDENCY_CMAKERC=ON"
   ];
 
-  postFixup = let
+
+  passAsFile = [ "vcpkgWrapper" ];
+  vcpkgWrapper = let
     # These are the most common binaries used by vcpkg
-    # Extra binaries can be added via overlay when needed
+    # Extra binaries can be added through override when needed
     runtimeDeps = [
       cacert
       cmake
+      curl
       git
       gzip
       meson
       ninja
       openssh
       python3
+      unzip
       zip
       zstd
     ] ++ extraRuntimeDeps;
+
+    # Apart from adding the runtime dependencies to $PATH,
+    # the wrapper will also override these arguments by default.
+    # This is to ensure that the executable does not try to
+    # write to the nix store. If the user tries to set any of the
+    # arguments themself, the wrapper will detect that the
+    # arguments are present, and prefer the user-provided value.
+    #
+    # It is also possible to override the cli arguments by
+    # settings either of the nix-specific environment variables.
+    argsWithDefault = [
+      {
+        arg = "--x-downloads-root";
+        env = "NIX_VCPKG_DOWNLOADS_ROOT";
+        default = "$NIX_VCPKG_WRITABLE_PATH/downloads";
+      }
+      {
+        arg = "--x-buildtrees-root";
+        env = "NIX_VCPKG_BUILDTREES_ROOT";
+        default = "$NIX_VCPKG_WRITABLE_PATH/buildtrees";
+      }
+      {
+        arg = "--x-packages-root";
+        env = "NIX_VCPKG_PACKAGES_ROOT";
+        default = "$NIX_VCPKG_WRITABLE_PATH/packages";
+      }
+      {
+        arg = "--x-install-root";
+        env = "NIX_VCPKG_INSTALL_ROOT";
+        default = "$NIX_VCPKG_WRITABLE_PATH/installed";
+      }
+    ];
   in ''
-    wrapProgram $out/bin/vcpkg --prefix PATH ${lib.makeBinPath runtimeDeps}
+    #!${runtimeShell}
+
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${XDG_CACHE_HOME+"$XDG_CACHE_HOME/vcpkg"}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${HOME+"$HOME/.vcpkg/root"}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${TMP}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-'/tmp'}
```

For someone who doesn't know well bash, this was not easy to grasp. What do you think of something like this? Longer, but more explicit
```suggestion
    get_vcpkg_path() {
      if [[ -n $NIX_VCPKG_WRITABLE_PATH ]]; then
          echo "$NIX_VCPKG_WRITABLE_PATH"
      elif [[ -n $XDG_CACHE_HOME ]]; then
          echo "$XDG_CACHE_HOME/vcpkg"
      elif [[ -n $HOME ]]; then
          echo "$HOME/.vcpkg/root"
      elif [[ -n $TMP ]]; then
          echo "$TMP"
      else
          echo "/tmp"
      fi
    }

    NIX_VCPKG_WRITABLE_PATH=$(get_vcpkg_path)
```